### PR TITLE
Testing with Mocks: User Repository

### PR DESCRIPTION
**Pull request by denisdimitrov**

### CODE REVIEW

File: src/userrepository/Validator.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package userrepository;
+
+/**
+ * Created by clouway on 16.08.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface Validator {
+    boolean validateAge(User user);
```
Comment (EmilGeorgiev) [inline src/userrepository/Validator.java]: Name of interface is common ```Validator``` and name of method is concrete ```validateAge()```, When name of interface is common, name of methods also should be common.

File: src/userrepository/AgeValidator.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ * Created by clouway on 16.08.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class AgeValidator implements Validator {
+    @Override
+    public boolean validateAge(User user) {
+        Integer age = Integer.valueOf(user.age);
+        return age >= 10 && age <= 100;
+    }
```
Comment (EmilGeorgiev) [inline src/userrepository/AgeValidator.java]:
```text
Забележка: Да се тестват методите в изолация. Да не се създават реални имплементации на валидатора и класа за връзка с базата данни. Задачата да се реализира с TDD подход.
```

File: src/userrepository/UserDB.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * Created by clouway on 16.08.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface UserDB {
+    boolean addUser(User user);
+
+    User checkUser(String name);
```
Comment (EmilGeorgiev) [inline src/userrepository/UserDB.java]: Boolean isAdult(String name) ;

File: src/userrepository/UserDB.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package userrepository;
+
+/**
+ * Created by clouway on 16.08.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public interface UserDB {
+    boolean addUser(User user);
```
Comment (EmilGeorgiev) [inline src/userrepository/UserDB.java]: add(User user)

File: src/userrepository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package userrepository;
+
+/**
+ * Created by clouway on 16.08.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class UserRepository {
```
Comment (EmilGeorgiev) [inline src/userrepository/UserRepository.java]: documentation

File: src/userrepository/UserRepository.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    public UserRepository(UserDB userDB, Validator validator) {
+        this.userDB = userDB;
+        this.validator = validator;
+    }
+
+    public void registeredUser(User user) {
+        if (validator.validateAge(user)) {
+            userDB.addUser(user);
+        }
+        throw new UserException("Cannot add user.");
```
Comment (EmilGeorgiev) [inline src/userrepository/UserRepository.java]:
```java
if !(validator.validate(user)) {
    throw new UserException("Cannot add user.");
}

userDB.addUser(user);
````

File: src/userrepository/UserRepository.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    public boolean isAdult(String name) {
+        User user = userDB.checkUser(name);
+        Integer age = Integer.valueOf(user.age);
```
Comment (EmilGeorgiev) [inline src/userrepository/UserRepository.java]: what will happen if user not exist in database

File: test/userrepositorytests/ValidatorTests.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Created by clouway on 16.08.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class ValidatorTests {
```
Comment (EmilGeorgiev) [inline test/userrepositorytests/ValidatorTests.java]: this test is unnecessary

File: src/userrepository/UserRepository.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    public final Validator validator;
+
+    public UserRepository(UserDB userDB, Validator validator) {
+        this.userDB = userDB;
+        this.validator = validator;
+    }
+
+    public void registeredUser(User user) {
+        if (validator.validateAge(user)) {
+            userDB.addUser(user);
```
Comment (EmilGeorgiev) [inline src/userrepository/UserRepository.java]: this is not tested

**Comment by denisdimitrov** [comment]
Everything is fixed.

File: src/userrepository/UserRepository.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * Checks if the user is in the database.
+     * If he is, it returns true/false.
+     * If he is not it throws UserException("message").
+     * @param name
+     * @return True if the user is above 18 years old. False if he is less than 18 years old.
+     */
+    public boolean isAdult(String name) {
+        User user = userDB.checkUser(name);
```
Comment (EmilGeorgiev) [inline src/userrepository/UserRepository.java]: userDB.findUser(name)

**Comment by EmilGeorgiev** [comment]
LGTM

**Pull request merged**